gh-120029: make `symtable.Symbol.__repr__` correctly reflect the compiler's flags

picnixz opened the pull request:
I did not expose `is_free_class` yet because I don't really know whether it's the best name or not, nor did I handle the `DEF_FREE` flag because I don't know its difference with the `FREE` scope itself (see https://github.com/python/cpython/issues/120029#issuecomment-2149236844).

I've nonethless cleaned up the lambda functions used to getting the scope ~~and also exported `SCOPE_OFFSET` instead of `SCOPE_OFF` (those constants are actually private since they are not documented so it should not affect public API). I can revert those changes easily though if you want me to make them in another PR.~~

cc @JelleZijlstra @carljm 

<!-- gh-issue-number: gh-120029 -->
* Issue: gh-120029
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--120099.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

picnixz committed:
export more flags

picnixz committed:
rename private module constant "SCOPE_OFF" -> "SCOPE_OFFSET"

picnixz committed:
add private Python function mirroring '_PyST_GetScope'

picnixz committed:
fix some reference in a comment

picnixz committed:
handle exported flags

picnixz committed:
add test

picnixz committed:
blurb

picnixz committed:
add doc

serhiy-storchaka commented on Lib/symtable.py:
```diff
@@ -311,6 +318,16 @@ def is_assigned(self):
+        """
+        return bool(self.__flags & DEF_COMP_ITER)
+
+    def is_comp_cell(self):
```
Is it the same as `is_comp_iter`?

picnixz replied:
Oups not at all! It's a typo, thanks for catching it!

serhiy-storchaka commented on Lib/symtable.py:
```diff
@@ -4,8 +4,11 @@
     USE,
     DEF_GLOBAL, DEF_NONLOCAL, DEF_LOCAL,
-    DEF_PARAM, DEF_TYPE_PARAM, DEF_IMPORT, DEF_BOUND, DEF_ANNOT,
-    SCOPE_OFF, SCOPE_MASK,
```
Renaming `SCOPE_OFF` is potentially a breaking change. Maybe add an alias just for the case?

picnixz replied:
Sure! Should I make it deprecated if someone imports it using the `__getattr__` module-level construction:

```python
def __getattr__(name):
    if name == 'SCOPE_OFF':
        import warnings
        warnings.warn(f"The {name!r} attribute is deprecated, use 'SCOPE_OFFSET' instead",
                      DeprecationWarning, stacklevel=2)
        return SCOPE_OFFSET

    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
```

serhiy-storchaka replied:
Since it is not documented, we are not obliged to keep it, so it can be removed at any time. But this can break the user code, and we want to be nice to users. If we are going to deprecate the old name, it is better to have several releases in which both names can be used without triggering a warning or error.

Alternatively, you can just remove this change from this PR. SCOPE_OFF was here for ages, and there is no great benefit in renaming it.

You can re-request whenever you want, but since I reviewed it only several hours ago, there is not much chance that I miss your changes and comments.

picnixz replied:
> But this can break the user code, and we want to be nice to users

Let's go with this mindset. I will also revert the change so that the code would at least be a bit nicer and the PR smaller. It's a bit weird that it was exported as `SCOPE_OFF` but let's keep what has been working for ages.

picnixz committed:
address review

- restore 'symtable.SCOPE_OFF' for compatibility purposes
- fix `Symbol.is_comp_cell()`

picnixz committed:
revert to using `SCOPE_OFF`

picnixz committed:
fixup new line

serhiy-storchaka reviewed (approved):
LGTM.

carljm commented on Lib/symtable.py:
```diff
@@ -4,7 +4,10 @@
     DEF_GLOBAL, DEF_NONLOCAL, DEF_LOCAL,
-    DEF_PARAM, DEF_TYPE_PARAM, DEF_IMPORT, DEF_BOUND, DEF_ANNOT,
+    DEF_PARAM, DEF_TYPE_PARAM,
+    DEF_FREE, DEF_FREE_CLASS,
```
I recommend leaving out `DEF_FREE` here. As you observed, it's never set on any name, and as I commented on the issue, I think we should consider removing it entirely. Its existence only serves to create confusion, since someone might expect it to be used.

At the very least I don't think we should further expose it.

carljm reviewed (changes requested):
`DEF_FREE_CLASS` is used in a case like this:

```
def f():
    x = 1
        class C:
            x = 2
            def method(self):
                return x
```

The `method()` method should return `1` -- it doesn't see the class-scoped `x`, but it does see the function-scoped `x` outside the class. The class-scoped `x` is marked `DEF_FREE_CLASS` in this case, to make that work.

It's such a niche use that I don't think we are going to find a method name here that nicely represents it in an intuitive way. Best option is probably just `is_free_class` to match the flag name.

picnixz committed:
expose ``is_free_class``

picnixz committed:
un-expose `DEF_FREE`

picnixz committed:
re-blurb

picnixz committed:
add test

picnixz committed:
fix lint

serhiy-storchaka commented on Doc/library/symtable.rst:
```diff
@@ -184,6 +184,25 @@ Examining Symbol Tables
+
+         def f():
+             x = 1  # function-scoped
+                 class C:
```
Wrong indentation.

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2024-06-05-11-03-10.gh-issue-120029.QBsw47.rst:
```diff
@@ -1,4 +1,4 @@
-Expose :meth:`symtable.Symbol.is_comp_iter` and :meth:`symtable.Symbol.is_comp_cell`
-in the :mod:`symtable` module.
+Expose :meth:`symtable.Symbol.is_free_class`, :meth:`symtable.Symbol.is_comp_iter`
```
A tip:  the resulting text will look better if write it as
```rst
Expose :class:`symtable.Symbol` methods :meth:`~symtable.Symbol.is_free_class`, :meth:`~symtable.Symbol.is_comp_iter`, ...
```
and remove ``` in the :mod:`symtable` module ``` because the module name is already included in the class reference.

Or keep ``` in the :mod:`symtable` module ```, but hide the module name in the class reference: ``` :class:`~symtable.Symbol` ```.

It does not matter much in the NEWS entry, but it would be more important in the What's New document. BTW, you can also add this in the What's New document.

picnixz committed:
fix indentation

picnixz committed:
update NEWS

picnixz committed:
update CHANGELOG

picnixz committed:
fix typo

picnixz committed:
fixup

JelleZijlstra commented on Doc/library/symtable.rst:
```diff
@@ -182,10 +184,41 @@ Examining Symbol Tables
 
+   .. method:: is_free_class()
+
+      Return *True* if a symbol is free from a method perspective.
```
```suggestion
      Return *True* if a class-scoped symbol is free from
      the perspective of a method.
```

picnixz committed:
address review

picnixz committed:
fixup

JelleZijlstra commented on Doc/library/symtable.rst:
```diff
@@ -182,10 +184,42 @@ Examining Symbol Tables
 
+   .. method:: is_free_class()
+
+      Return *True* if a class-scoped symbol is free from
```
Though strictly speaking this can also get used for symbols in annotation scopes:

```
def f():
    x = 1  # function-scoped
    class C:
        x = 2
        type Alias = (x, lambda: x)
```


```
            symbol table for nested type alias 'Alias':
                free symbol '__classdict__': use
                global_implicit symbol 'x': use, def_free_class

                symbol table for nested function 'lambda':
                    free symbol 'x': use
```

But I think it's still fair to say "class-scoped symbol" here, as the symbol is inherited from the class scope.

picnixz replied:
If we go into that direction, I don't know how to name that 'x' inside... "class-scoped" seemed clear enough IMO and if we add something like "inherited class-scoped symbols in annotation scopes", I feel it could be more confusing (unless we add an example).

carljm committed:
Merge branch 'main' into export-symtable-flags

picnixz committed:
Merge branch 'main' into export-symtable-flags

picnixz committed:
fix test

serhiy-storchaka commented on Lib/test/test_symtable.py:
```diff
@@ -319,7 +319,8 @@ def test_symbol_repr(self):
                                 "?", "exec")
-        func_f = st3.get_children()[0]
+        # child 0 is for __annotate__
+        func_f = st3.get_children()[1]
```
Maybe use `lookup()` instead of `get_children()`?

picnixz replied:
Using `lookup` gives me an instance of `Symbol` but I want the symbol table of 'f' and then get the symbol table of 'A' and finally lookup for the symbol 'x'. 

It's something that I wanted to ask as well, but would it make sense to have a better find interface where we could also find children instead of getting them as a list?

(technically, I could still get the corresponding symbol table from the symbol itself but I don't think `st3.lookup('f').get_namespace()` is better)

serhiy-storchaka replied:
Oh, I see. I think that the current interface is fine, taking into account that there may be several symbol tables with the same name, so we would need to add methods for combinations `lookup()`+`get_namespace()` and `lookup()`+`get_namespaces()`.

picnixz replied:
Yes, that's why I was a bit conflicted about whether implementing a better find function or not. It's a bit annoying to use `get_children()[i]` but it's only for the tests so I can live with that.
